Add bit-banging based driver for WS2812B LED strip

jhedberg opened the pull request:
These two patches have been tested with the ZIP Halo LED strip for the BBC micro:bit:

https://www.kitronik.co.uk/5625-zip-halo-for-the-bbc-microbit.html

The implementation utilizes inline assembly and only supports the Cortex-M0 for now.

Commit:
drivers: led_strip: Add driver for software-based WS2812B operation

This driver uses a bit-banging based technique of generating a signal
for the WS2812B LED strip. Since bit-banging is very timing sensitive,
where each CPU cycle counts, the driver uses inline assembly to
perform the most critical operataions. This initial version of the
driver only supports a Cortex-M0 implementation, and can e.g. be used
with the ZIP Halo LED strip for the BBC microbit:

https://www.kitronik.co.uk/5625-zip-halo-for-the-bbc-microbit.html

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
Note: this driver does not yet support coexistence with Bluetooth. That's the next step I'm working on - it'll involve integration with the Bluetooth controller's ticker module.

lpereira commented on samples/drivers/led_ws2812b_sw/README.rst:
```diff
@@ -0,0 +1,42 @@
+.. _led_ws2812b_sw_sample:
+
+WS2812B Sample Application
+######$###################
```
Stray `$`

jhedberg replied:
fixed, thanks

dbkinder reviewed (approved):
thanks!

lpereira commented on samples/drivers/led_ws2812b_sw/src/main.c:
```diff
@@ -0,0 +1,82 @@
+	struct device *strip;
+	size_t i, time;
+
+	strip = device_get_binding(STRIP_DEV_NAME);
```
Can't the `led_ws2812` sample be used instead, just not building the call/check to `device_get_binding(SPI_DEV_NAME)`? Apart from that (and the number of LEDs), the samples are identical.

jhedberg replied:
Yes, I could add some #ifdefs to the existing sample instead of adding a new one.

jhedberg replied:
I've completed the changes locally, but I noticed that this breaks the pattern a little bit: it seems all LED drivers have essentially a copy of the same sample if you look at `samples/drivers/led_*`

jhedberg replied:
I.e. if we start adding `#ifdefs` maybe all drivers should be collected into the same sample (however in that case the name of the sample starts being misleading since it's not ws2812 anymore)

jhedberg replied:
I've pushed now an extension of the existing sample instead of adding a new one. Let me know how that looks. I also had to remove any SPI & ws2812 driver stuff from prj.conf and move them to the board/*.conf.

cvinayak commented on drivers/led_strip/Kconfig:
```diff
@@ -46,6 +46,8 @@ source "drivers/led_strip/Kconfig.lpd880x"
 
 source "drivers/led_strip/Kconfig.ws2812"
 
+source "drivers/led_strip/Kconfig.ws2812b_sw"
```
update copyright on this file?

jhedberg replied:
done

cvinayak commented on drivers/led_strip/Kconfig.ws2812b_sw:
```diff
@@ -0,0 +1,38 @@
+
+menuconfig WS2812B_SW
+	bool "Enable WS2812B software-based LED strip driver"
+	# Only a Cortex-M0 inline assembly implementation is supported for now
```
Is the implementation dependent on CPU clock speed? If so, state it or we need a way for the driver to be more portable across CPU speeds.

jhedberg replied:
Yes, it's dependent on the speed, and I wouldn't expect each CPU/architecture to have their own inline assembly snippet. I'll add a comment to state the requirement, but I think it should be in the code itself rather than here in the Kconfig.

jhedberg replied:
Just another note regarding CPU speed, based on what I've found when doing research online, it seems the minimum theoretical CPU speed that should be able to accomplish the signal generation task is about 4 MHz.

cvinayak commented on drivers/led_strip/ws2812b_sw.c:
```diff
@@ -0,0 +1,141 @@
+static int send_buf(u8_t *buf, size_t len)
+{
+	/* Address of OUTSET. OUTCLR is OUTSET + 4 */
+	volatile u32_t *base = (u32_t *)(NRF_GPIO_BASE + 0x508);
```
This is very NRF5 specific, do you want to conditional compile? or have depends on?

Thinking aloud, @dbkinder @SebastianBoe is there a build target that can list features/drivers and corresponding supported SoC/Board in Zephyr?
Example, make help -DFEATURES  or something

jhedberg replied:
The Kconfig entry has a `depends on CPU_CORTEX_M0`. Is that sufficient? I could add a pre-processor `#error` to the code as well in case the architecture isn't supported, however that'd be a bit redundant with the Kconfig dependency.

jhedberg replied:
Ah, right, CORTEX_M0 doesn't mean it's nRF, of course. I'll fix this, i.e. change the dependency to be for nRF51 for now.

jhedberg replied:
For the record, here's an example of multi-arch #ifdefs that another implementation uses:

https://github.com/adafruit/Adafruit_NeoPixel/blob/master/Adafruit_NeoPixel.cpp

However, I hope we can do something a bit cleaner than that.

cvinayak replied:
Could you have separate assembly files please, when you go for many variants. This inline asm in c is hard on eyes!

Commit:
samples: led_strip: Extend ws2812 sample with ws2812b_sw support

Add support for building the WS2812 LED strip driver sample against
the new ws2812b_sw driver. Currently a configuration is only provided
for the BBC micro:bit board.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

cvinayak reviewed (approved):
LGTM

mbolivar commented on drivers/led_strip/ws2812b_sw.c:
```diff
@@ -0,0 +1,144 @@
+		pixels[i].g = r;
+	}
+
+	return send_buf((u8_t *)pixels, num_pixels * sizeof(*pixels));
```
This cast doesn't work if CONFIG_LED_STRIP_RGB_SCRATCH=y (which could conceivably happen if you want to use an APA102 on the same board). Just to be safe, can you be a bit more defensive?

jhedberg replied:
I've added `depends on !CONFIG_LED_STRIP_RGB_SCRATCH` to Kconfig to protect against this happening.

This actually brings me to another efficiency related issue with the current LED API: it'd be nice if we could find some solution which would let the application pass const (ROM) buffers to the API, and have those drivers which don't have the scratch u8_t be able to pass the buffer as-is to the hardware. This wouldn't just be about avoiding translations in the driver, but allowing e.g. bigger animations stored in ROM when there's not enough space in RAM for them.

mbolivar replied:
> I've added depends on !CONFIG_LED_STRIP_RGB_SCRATCH to Kconfig to protect against this happening.

Hrm... why not just get it right for both cases? I don't think it's going to blow any resource budgets to do so. I also note that you're relying on the order of r/g/b in the struct earlier in this routine, which also seems unnecessary.

What's wrong with the approach taken by lpd880x_strip_update_rgb(), which makes no assumptions about struct layout or whether scratch is present?

> This actually brings me to another efficiency related issue with the current LED API: it'd be nice if we could find some solution which would let the application pass const (ROM) buffers to the API, and have those drivers which don't have the scratch u8_t be able to pass the buffer as-is to the hardware. This wouldn't just be about avoiding translations in the driver, but allowing e.g. bigger animations stored in ROM when there's not enough space in RAM for them.

I agree this would be very nice, and that the current API is RAM-hungry in this regard. It's possibly tricky to do in general, though. For example, the SPI-based ws2812 driver would require an 8x storage increase to store a "raw" array of SPI data, since it uses a one byte SPI frame for every bit of pixel data.

Do you have any ideas for a general mechanism? Otherwise, perhaps an ioctl()-style interface where individual drivers can specify chip-specific extensions could perhaps serve.

jhedberg commented:
>  Hrm... why not just get it right for both cases? I don't think it's going to blow any resource budgets to do so. I also note that you're relying on the order of r/g/b in the struct earlier in this routine, which also seems unnecessary.

> What's wrong with the approach taken by lpd880x_strip_update_rgb(), which makes no assumptions about struct layout or whether scratch is present?

That's a fair point. Since I'm having to mess with the given buffer anyway, re-encoding all three pixels is not that much overhead. I've pushed a version that does this (and removes the `depends on` I added earlier).

jhedberg commented:
@mbolivar are you ok with the changes I made? Your review seems to be the only thing preventing this one from being merged.

jhedberg commented:
> @mbolivar are you ok with the changes I made? Your review seems to be the only thing preventing this one from being merged.

@mbolivar ping

mbolivar reviewed (approved):
Looks good, thanks for the update
